#### Importing necessary libraries

In [117]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

#### Loading the dataset

  Got this error: DtypeWarning: Columns (76) have mixed types. Specify dtype option on import or set low_memory=False.
  So used : low_memory as False 

In [118]:
df = pd.read_csv('fifa21_raw_data.csv', low_memory=False)

#### Adjusting pandas display options to show all columns and rows

In [119]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [120]:
df.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   
2          Right   91            GK  Jul 16, 2014           NaN    €120M   
3          Right   91           CAM  Aug 30, 2015           NaN    €129M   
4          Right   91            LW   Aug 3, 2017           NaN    €132M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   
2  €125K        €159.4M         95        13         11                15   
3  €370K          €161M        407        94         82                55   
4  €270K        €166.5M        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioni

In [121]:
df.tail()

ID          Name        LongName  \
18974  247223        Xia Ao          Ao Xia   
18975  258760      B. Hough       Ben Hough   
18976  252757   R. McKinley  Ronan McKinley   
18977  243790  Wang Zhen'ao    Zhen'ao Wang   
18978  252520     Zhou Xiao       Xiao Zhou   

                                               photoUrl  \
18974  https://cdn.sofifa.com/players/247/223/21_60.png   
18975  https://cdn.sofifa.com/players/258/760/21_60.png   
18976  https://cdn.sofifa.com/players/252/757/21_60.png   
18977  https://cdn.sofifa.com/players/243/790/21_60.png   
18978  https://cdn.sofifa.com/players/252/520/21_60.png   

                                               playerUrl Nationality  Age  \
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR   21   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England   17   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England   18   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR   20   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR   21   

       ↓OVA  POT                      Club     Contract Positions Height  \
18974    47   55        \n\n\n\nWuhan Zall  2018 ~ 2022        CB  178cm   
18975    47   67   \n\n\n\nOldham Athletic  2020 ~ 2021        CM  175cm   
18976    47   65        \n\n\n\nDerry City  2019 ~ 2020        CM  179cm   
18977    47   57  \n\n\n\nDalian YiFang FC  2020 ~ 2022        RW  175cm   
18978    47   57  \n\n\n\nDalian YiFang FC  2019 ~ 2023    CB, LB  188cm   

      Weight Preferred Foot  BOV Best Position        Joined Loan Date End  \
18974   66kg          Right   49            CB  Jul 13, 2018           NaN   
18975   65kg          Right   51           CAM   Aug 1, 2020           NaN   
18976   74kg          Right   49           CAM   Mar 8, 2019           NaN   
18977   69kg          Right   48            ST  Sep 22, 2020           NaN   
18978   75kg           Left   50            LB  Jul 29, 2019           NaN   

       Value  Wage Release Clause  Attacking  Crossing  Finishing  \
18974  €100K   €1K           €70K        145        23         26   
18975  €130K  €500          €165K        211        38         42   
18976  €120K  €500          €131K        200        30         34   
18977  €100K   €2K           €88K        215        45         52   
18978  €100K   €1K           €79K        163        40         18   

       Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
18974                43             26       27    142         27     23   
18975                40             56       35    219         46     40   
18976                43             54       39    207         43     39   
18977                34             42       42    194         51     35   
18978                40             45       20    171         40     33   

       FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
18974           21            29            42       294            68   
18975           35            50            48       305            63   
18976           31            47            47       290            59   
18977           31            31            46       254            62   
18978           20            40            38       279            60   

       Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
18974            60       69         46       51    221          36       57   
18975            64       61         51       66    226          48       58   
18976            66       51         47       67    242          45       52   
18977            55       50         33       54    235          56       45   
18978            63       57         47       52    231          37       66   

       Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
18974       54        50          24        192          48             50   
18975       43        47      

In [122]:
# Find duplicate rows
duplicate_rows = df[df.duplicated()]

# Display duplicate rows
print("Duplicate Rows:")
print(duplicate_rows)

Duplicate Rows:
Empty DataFrame
Columns: [ID, Name, LongName, photoUrl, playerUrl, Nationality, Age, ↓OVA, POT, Club, Contract, Positions, Height, Weight, Preferred Foot, BOV, Best Position, Joined, Loan Date End, Value, Wage, Release Clause, Attacking, Crossing, Finishing, Heading Accuracy, Short Passing, Volleys, Skill, Dribbling, Curve, FK Accuracy, Long Passing, Ball Control, Movement, Acceleration, Sprint Speed, Agility, Reactions, Balance, Power, Shot Power, Jumping, Stamina, Strength, Long Shots, Mentality, Aggression, Interceptions, Positioning, Vision, Penalties, Composure, Defending, Marking, Standing Tackle, Sliding Tackle, Goalkeeping, GK Diving, GK Handling, GK Kicking, GK Positioning, GK Reflexes, Total Stats, Base Stats, W/F, SM, A/W, D/W, IR, PAC, SHO, PAS, DRI, DEF, PHY, Hits]
Index: []


In [123]:
# Check for missing values
missing_values = df.isnull().sum()

# Display missing values count
print("Missing Values:")
print(missing_values)

Missing Values:
ID                      0
Name                    0
LongName                0
photoUrl                0
playerUrl               0
Nationality             0
Age                     0
↓OVA                    0
POT                     0
Club                    0
Contract                0
Positions               0
Height                  0
Weight                  0
Preferred Foot          0
BOV                     0
Best Position           0
Joined                  0
Loan Date End       17966
Value                   0
Wage                    0
Release Clause          0
Attacking               0
Crossing                0
Finishing               0
Heading Accuracy        0
Short Passing           0
Volleys                 0
Skill                   0
Dribbling               0
Curve                   0
FK Accuracy             0
Long Passing            0
Ball Control            0
Movement                0
Acceleration            0
Sprint Speed            0
Agility               

From the missing values summary provided, we can draw the following conclusions:

1. **Loan Date End**: There are 17,966 missing values in the "Loan Date End" column. This suggests that a significant number of players in the dataset are not currently on loan, as indicated by the missing values in this column.

2. **Hits**: There are 2,595 missing values in the "Hits" column. This column likely represents the number of hits or views on a player's profile, and the missing values could indicate that data for this metric is not available for all players.

Overall, the missing values in the dataset do not seem to significantly affect the majority of the columns, except for "Loan Date End" and "Hits". Depending on the analysis or modeling tasks, these missing values may need to be handled appropriately, such as through imputation or removal. Additionally, further investigation into the reasons for missing values in these columns may provide valuable insights into the data collection process or the characteristics of the players in the dataset.

In [124]:
# Print all data types
print(df.dtypes)

ID                   int64
Name                object
LongName            object
photoUrl            object
playerUrl           object
Nationality         object
Age                  int64
↓OVA                 int64
POT                  int64
Club                object
Contract            object
Positions           object
Height              object
Weight              object
Preferred Foot      object
BOV                  int64
Best Position       object
Joined              object
Loan Date End       object
Value               object
Wage                object
Release Clause      object
Attacking            int64
Crossing             int64
Finishing            int64
Heading Accuracy     int64
Short Passing        int64
Volleys              int64
Skill                int64
Dribbling            int64
Curve                int64
FK Accuracy          int64
Long Passing         int64
Ball Control         int64
Movement             int64
Acceleration         int64
Sprint Speed         int64
A

**FIFA 21 Dataset Column Descriptions**

| Column Name | Data Type  | Description |
|---|---|---| 
| ID | `int64` | Unique identifier for the player |
| Name | `object` | Player's short name |
| LongName | `object` | Player's full name |
| photoUrl | `object` | URL to the player's profile picture |
| playerUrl | `object` | URL to the player's full profile |
| Nationality | `object` | Player's country of origin |
| Age | `int64` | Player's age |
| ↓OVA | `int64` |  Player's overall rating|
| POT | `int64` | Player's potential rating| 
| Club | `object` | Player's current club |
| Contract | `object` | Details of the player's contract |
| Positions | `object` | Player's positions on the field |
| Height | `object` | Player's height |
| Weight | `object` | Player's weight |
| Preferred Foot | `object` | Player's preferred foot (left or right) |
| BOV | `int64` | Best overall rating of the player |
| Best Position | `object` | Player's best position on the field |
| Joined | `object` | Date when the player joined the current club |
| Loan Date End | `object` | End date of the player's loan (if applicable) |
| Value | `object` | Current transfer market value of the player |
| Wage | `object` | Weekly wage of the player |
| Release Clause | `object` | Release clause amount in the player's contract |
| Attacking | `int64` | Overall rating of the player's attacking abilities |
| Crossing | `int64` | Player's skill level in crossing |
| Finishing | `int64` | Player's skill level in finishing |
| Heading Accuracy | `int64` | Player's skill level in heading accuracy |
| Short Passing | `int64` | Player's skill level in short passing |
| Volleys | `int64` | Player's skill level in volleys |
| Skill | `int64` | Overall rating of the player's skill moves |
| Dribbling | `int64` | Player's skill level in dribbling |
| Curve | `int64` | Player's skill level in curve shots |
| FK Accuracy | `int64` | Player's skill level in free-kick accuracy |
| Long Passing | `int64` | Player's skill level in long passing |
| Ball Control | `int64` | Player's skill level in ball control |
| Movement | `int64` | Overall rating of the player's movement attributes |
| Acceleration | `int64` | Player's acceleration skill level |
| Sprint Speed | `int64` | Player's sprint speed skill level |
| Agility | `int64` | Player's agility skill level |
| Reactions | `int64` | Player's reaction skill level|
| Balance | `int64` | Player's balance skill level |
| Power | `int64` | Overall rating of the player's power attributes |
| Shot Power | `int64` | Player's shot power skill level |
| Jumping | `int64` | Player's jumping skill level|
| Stamina | `int64` | Player's stamina skill level|
| Strength | `int64` | Player's strength skill level |
| Long Shots | `int64` | Player's long shots skill level |
| Mentality | `int64` | Overall rating of the player's mental attributes |
| Aggression | `int64` | Player's aggression skill level |
| Interceptions | `int64` | Player's interceptions skill level |
| Positioning | `int64` | Player's positioning skill level|
| Vision | `int64` | Player's vision skill level |
| Penalties | `int64` | Player's penalty kicks skill level |
| Composure | `int64` | Player's composure skill level |
| Defending | `int64` | Overall rating of the player's defending attributes |
| Marking | `int64` | Player's marking skill level|
| Standing Tackle | `int64` | Player's standing tackle skill level |
| Sliding Tackle | `int64` | Player's sliding tackle skill level|
| Goalkeeping | `int64` | Overall rating of the player's goalkeeping attributes |
| GK Diving | `int64` | Player's diving skill level (goalkeeping)|
| GK Handling | `int64` | Player's handling skill level (goalkeeping)|
| GK Kicking | `int64` | Player's kicking skill level (goalkeeping) |
| GK Positioning | `int64` | Player's positioning skill level (goalkeeping)|
| GK Reflexes | `int64` | Player's reflexes skill level (goalkeeping)|
| Total Stats | `int64` | Total statistics of the player |
| Base Stats | `int64` | Base statistics of the player |
| W/F | `object` | Player's weak foot skill level |
| SM | `object` | Player's skill moves level |
| A/W | `object` | Attacking work rate of the player |
| D/W | `object` | Defensive work rate of the player |
| IR | `object` | International reputation of the player |
| PAC | `int64` | Player's pace attribute |
| SHO | `int64` | Player's shooting attribute |
| PAS | `int64` | Player's passing attribute |
| DRI | `int64` | Player's dribbling attribute |
| DEF | `int64` | Player's defending attribute |
| PHY | `int64` | Player's physical attribute |
| Hits | `object` | Number of hits |
